jupyterlab_git/git.py

```diff
@@ -3,61 +3,111 @@
+    
+    Args:
+        cmdline (List[str]): Command line to be executed
+        username (str): User name
+        password (str): User password
+        cwd (Optional[str]): Current working directory
+        env (Optional[Dict[str, str]]): Defines the environment variables for the new process.
+    Returns:
+        (int, str): (return code, output)
     """
-    def __init__(self, command, cwd, env, username, password):
-        self.command = command
-        self.cwd = cwd
-        self.env = env
-        self.username = username
-        self.password = password
-
-    def communicate(self):
+
+    def call_subprocess(
+        cmdline: "List[str]",
+        username: "str",
+        password: "str",
+        cwd: "Optional[str]" = None,
+        env: "Optional[Dict[str, str]]" = None,
+    ) -> "Tuple[int, str]":
         try:
             p = pexpect.spawn(
-                self.command, 
-                cwd = self.cwd,
-                env = self.env
+                cmdline[0], cmdline[1:], cwd=cwd, env=env, encoding="utf-8"
             )
 
             # We expect a prompt from git
             # In most of cases git will prompt for username and
             #  then for password
             # In some cases (Bitbucket) username is included in
-            #  remote URL, so git will not ask for username 
-            i = p.expect(['Username for .*: ', 'Password for .*:'])
-            if i==0: #ask for username then password
-                p.sendline(self.username)
-                p.expect('Password for .*:')
-                p.sendline(self.password)
-            elif i==1: #only ask for password
-                p.sendline(self.password)
+            #  remote URL, so git will not ask for username
+            i = p.expect(["Username for .*: ", "Password for .*:"])
+            if i == 0:  # ask for username then password
+                p.sendline(username)
+                p.expect("Password for .*:")
+                p.sendline(password)
+            elif i == 1:  # only ask for password
+                p.sendline(password)
 
             p.expect(pexpect.EOF)
             response = p.before
 
-            self.returncode = p.wait()
+            returncode = p.wait()
             p.close()
-            
-            return response
-        except pexpect.exceptions.EOF: #In case of pexpect failure
+
+            return returncode, response
+        except pexpect.exceptions.EOF:  # In case of pexpect failure
             response = p.before
-            self.returncode = p.exitstatus
-            p.close() #close process
-            return response
+            returncode = p.exitstatus
+            p.close()  # close process
+            return returncode, response
+
+    current_loop = tornado.ioloop.IOLoop.current()
+    returncode, output = await current_loop.run_in_executor(
+        None, call_subprocess, cmdline, username, password, cwd, env
+    )
+
+    return returncode, output
```

Comment:
I don't love that the return types of `execute` and `execute_with_authentication` are different. Do they have to be?

Reply:
I fully agree (I started to work on that actually 😉 ). But `pexpect.spawn` is by definition unable to split stdout and stderr.

Looking around, [`pexpect.fdpexpect.fdspawn`](https://pexpect.readthedocs.io/en/stable/api/fdpexpect.html) could work (see [that discussion](https://stackoverflow.com/questions/27179383/extracting-stderr-from-pexpect)). It has the advantage that it should address #492 and be consistent with `subprocess.Popen`.

@telamonian @ktaletsk should we move to `fdspawn` to have a stream separation (and to support Windows) ?